runtime_scan/pkg/provider/docker/client.go

```diff
@@ -0,0 +1,585 @@
+// Copyright © 2023 Cisco Systems, Inc. and its affiliates.
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package docker
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	types "github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/api/types/mount"
+	"github.com/docker/docker/api/types/volume"
+	"github.com/docker/docker/client"
+	"time"
+
+	"github.com/openclarity/vmclarity/api/models"
+	"github.com/openclarity/vmclarity/runtime_scan/pkg/provider"
+	"github.com/openclarity/vmclarity/shared/pkg/utils"
+	"io"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"sync"
+)
+
+type Client struct {
+	dockerClient *client.Client
+	config       *Config
+}
+
+func New(ctx context.Context) (*Client, error) {
+	config, err := NewConfig()
+	if err != nil {
+		return nil, fmt.Errorf("invalid configuration. Provider=%s: %w", models.Docker, err)
+	}
+
+	if err = config.Validate(); err != nil {
+		return nil, fmt.Errorf("failed to validate provider configuration. Provider=%s: %w", models.Docker, err)
+	}
+
+	dockerClient, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		return nil, fmt.Errorf("failed to load provider configuration. Provider=%s: %w", models.Docker, err)
+	}
+
+	return &Client{
+		dockerClient: dockerClient,
+		config:       config,
+	}, nil
+}
+
+func (c Client) Kind() models.CloudProvider {
+	return models.Docker
+}
+
+func (c Client) DiscoverAssets(ctx context.Context) ([]models.AssetType, error) {
+	var err error
+	wg := &sync.WaitGroup{}
+	errs := make(chan error, 2)
+	assets := make(chan models.AssetType)
+
+	wg.Add(1)
+	go c.getImages(ctx, assets, errs, wg)
+
+	wg.Add(1)
+	go c.getContainers(ctx, assets, errs, wg)
+
+	go func() {
+		wg.Wait()
+		close(errs)
+		close(assets)
+	}()
+
+	var ret []models.AssetType
+	for t := range assets {
+		ret = append(ret, t)
+	}
+
+	for e := range errs {
+		if e != nil {
+			// nolint:typecheck
+			err = errors.Join(err, e)
+		}
+	}
+	if err != nil {
+		return nil, err
+	}
```

Comment:
The assumption that there will only be two errors will lead to a deadlock. Note that both methods `getImages` and `getContainers` write to assets and errors channels multiple times (for example, in the loop). When we have a situation in which 3 errors have been written to a channel and asset channel consumer is still waiting, this will block.

Reply:
I also think that there is no need to call getImages and getContainers in a spearate gorotine if these two will perform the requests concurrently to each image/container internally. Fetching the list of containers/images is a relatively minor operation compared to batch processing all of them.